File: leetcode/3sum.cc
// https://leetcode.com/problems/3sum/

// time complexity:

#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int> &nums) {
        int len = nums.size();
        if (len < 3)
            return {};

        vector<vector<int>> sums;
        sort(nums.begin(), nums.end());
        for (int i = 0, left, right, sum; i < len - 2; i++) {
            if (i > 0 && nums[i - 1] == nums[i])
                continue;

            left = i + 1, right = len - 1;
            while (left < right) {
                sum = nums[i] + nums[left] + nums[right];
                if (sum < 0)
                    left++;
                else if (sum > 0)
                    right--;
                else {
                    sums.push_back({nums[i], nums[left++], nums[right--]});
                    while (left < right && nums[left - 1] == nums[left])
                        left++;
                    while (left < right && nums[right] == nums[right + 1])
                        right--;
                }
            }
        }
        return sums;
    }
};

File: leetcode/3sum.py
# https://leetcode.com/problems/3sum/

# time complexity: 

class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        triplets = []
        if len(nums) < 3:
            return triplets
        nums.sort()
        for i in range(len(nums) - 2):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            j, k = i + 1, len(nums) - 1
            while j < k:
                if (sum := nums[i] + nums[j] + nums[k]) == 0:
                    triplets.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while j < k and nums[k - 1] == nums[k]:
                        k -= 1
                    j += 1
                    k -= 1
                elif sum < 0:
                    j += 1
                else:
                    k -= 1
        return triplets